x/superfluid/keeper/migrate.go

```diff
@@ -155,160 +156,157 @@ func (k Keeper) migrateSuperfluidUnbondingBalancerToConcentrated(ctx sdk.Context
+	// It also returns the lock object that contains the remaining shares that were not used in this migration.
+	exitCoins, remainingSharesLock, err := k.validateSharesToMigrateUnlockAndExitBalancerPool(ctx, sender, poolIdLeaving, preMigrationLock, sharesToMigrate, tokenOutMins, remainingLockTime)
 	if err != nil {
-		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
 	}
 
 	// Create a new lock that is unlocking for the remaining time of the old lock.
 	// Regardless of the previous lock's status, we create a new lock that is unlocking.
+	// This is because locking without superfluid is pointless in the context of concentrated liquidity.
 	positionId, amount0, amount1, liquidity, joinTime, concentratedLockId, err = k.clk.CreateFullRangePositionUnlocking(ctx, poolIdEntering, sender, exitCoins, remainingLockTime)
 	if err != nil {
-		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+		return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
 	}
 
 	// If there are remaining gamm shares, we must re-lock them.
-	remainingGammShares := gammSharesInLock.Sub(sharesToMigrate)
-	if !remainingGammShares.IsZero() {
-		// Create a new lock with the remaining gamm shares for the remaining lock time.
-		newLock, err := k.lk.CreateLock(ctx, sender, sdk.NewCoins(remainingGammShares), remainingLockTime)
-		gammLockId = newLock.ID
-		if err != nil {
-			return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
-		}
-
+	if remainingSharesLock.ID != 0 {
+		gammLockId = remainingSharesLock.ID
 		// If the gamm lock was unlocking, we begin the unlock from where it left off.
 		if wasUnlocking {
-			_, err := k.lk.BeginForceUnlock(ctx, newLock.ID, newLock.Coins)
+			_, err := k.lk.BeginForceUnlock(ctx, remainingSharesLock.ID, remainingSharesLock.Coins)
 			if err != nil {
-				return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, err
+				return 0, sdk.Int{}, sdk.Int{}, sdk.Dec{}, time.Time{}, 0, 0, 0, 0, err
 			}
 		}
 	}
 
-	return positionId, amount0, amount1, liquidity, joinTime, gammLockId, concentratedLockId, nil
+	return positionId, amount0, amount1, liquidity, joinTime, gammLockId, concentratedLockId, poolIdLeaving, poolIdEntering, nil
 }
 
-// prepareMigration prepares for the migration of gamm LP tokens from the Balancer pool to the Concentrated pool. It performs the following steps:
+// routeMigration determines the status of the provided lock which is used to determine the method for migration.
+// It also returns the underlying synthetic locks of the provided lock, if any exist.
+func (k Keeper) routeMigration(ctx sdk.Context, sender sdk.AccAddress, lockId uint64, sharesToMigrate sdk.Coin) (synthLocksBeforeMigration []lockuptypes.SyntheticLock, migrationType MigrationType, err error) {
```

Comment:
Can we have unit test for this method?

Reply:
Each path is tested in each individual method test:

https://github.com/osmosis-labs/osmosis/blob/058fcc43f9ae8872a97aa2350f9aa58d9c3861b9/x/superfluid/keeper/migrate_test.go#L244-L246

https://github.com/osmosis-labs/osmosis/blob/058fcc43f9ae8872a97aa2350f9aa58d9c3861b9/x/superfluid/keeper/migrate_test.go#L363-L366

https://github.com/osmosis-labs/osmosis/blob/058fcc43f9ae8872a97aa2350f9aa58d9c3861b9/x/superfluid/keeper/migrate_test.go#L454-L458